src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -196,56 +198,70 @@ private[concurrent] trait BatchingExecutor extends Executor {
-        reportFailure(failure)
-        runWithoutResubmit(runUntilFailureOrDone())
-      } else {
-        _tasksLocal.set(BatchingExecutorStatics.marker)
-        failure
+    @tailrec override final def run(): Unit = {
+      try runN(BatchingExecutorStatics.runLimit) catch {
+        case ie: InterruptedException =>
+          reportFailure(ie) // TODO: Handle InterruptedException differently?
+        case f if NonFatal(f) =>
+          reportFailure(f)
       }
 
-    override final def run(): Unit = {
-      _tasksLocal.set(this) // This is later cleared in `runWithoutResubmit`
-
-      val f = runWithoutResubmit(runUntilFailureOrDone())
-
-      if (f != null)
-        throw f
+      if (this.size > 0)
+        run()
     }
   }
 
-  /** SHOULD throw a NullPointerException when `runnable` is null
+  /** MUST throw a NullPointerException when `runnable` is null
+   * When implementing a sync BatchingExecutor, it is RECOMMENDED
+   * to implement this method as `runnable.run()`
   */
-  protected def submitAsync(runnable: Runnable): Unit
+  protected def submitForExecution(runnable: Runnable): Unit
 
   /** Reports that an asynchronous computation failed.
    *  See `ExecutionContext.reportFailure(throwable: Throwable)`
   */
   protected def reportFailure(throwable: Throwable): Unit
 
+  /**
+   * WARNING: Never use both `submitAsyncBatched` and `submitSyncBatched` in the same
+   * implementation of `BatchingExecutor`
+   */
   protected final def submitAsyncBatched(runnable: Runnable): Unit = {
     val b = _tasksLocal.get
     if (b.isInstanceOf[AsyncBatch]) b.asInstanceOf[AsyncBatch].push(runnable)
-    else submitAsync(new AsyncBatch(runnable))
+    else submitForExecution(new AsyncBatch(runnable))
   }
 
+  /**
+   * WARNING: Never use both `submitAsyncBatched` and `submitSyncBatched` in the same
+   * implementation of `BatchingExecutor`
+   */
   protected final def submitSyncBatched(runnable: Runnable): Unit = {
     Objects.requireNonNull(runnable, "runnable is null")
-    val b = _tasksLocal.get
+    val tl = _tasksLocal
+    val b = tl.get
     if (b.isInstanceOf[SyncBatch]) b.asInstanceOf[SyncBatch].push(runnable)
-    else if (b == null) {                             // If there is null in _tasksLocal, set a marker and run, inflate the Batch only if needed
-      _tasksLocal.set(BatchingExecutorStatics.marker) // Set a marker to indicate that we are submitting synchronously
-      runnable.run()                                  // If we observe a non-null task which isn't a batch here, then allocate a batch
-      _tasksLocal.remove()                            // Since we are executing synchronously, we can clear this at the end of execution
-    } else new SyncBatch(runnable).run()
+    else {
+      val i = if (b ne null) b.asInstanceOf[java.lang.Integer].intValue else 0
+      if (i < BatchingExecutorStatics.syncPreBatchDepth) {
+        tl.set(java.lang.Integer.valueOf(i + 1))
```

Comment:
We use j.l.Integer here to control the bytecode generated—the Integers we will be using here are all internally cached inside Integer, so no allocations on this path.